plugins/net_plugin/net_plugin.cpp

```diff
@@ -2182,14 +2184,38 @@ namespace eosio {
               sync_next_expected_num < sync_last_requested_num );
    }
 
+   // called from c's connection strand
    bool sync_manager::is_sync_request_ahead_allowed(block_num_type blk_num) const REQUIRES(sync_mtx) {
       if (blk_num >= sync_last_requested_num) {
          // do not allow to get too far ahead (sync_fetch_span) of chain head
          // use chain head instead of fork head so we do not get too far ahead of applied blocks
-         uint32_t head = my_impl->get_chain_head_num();
-         if (blk_num < head + sync_fetch_span)
+         uint32_t head_num = my_impl->get_chain_head_num();
+         if (blk_num < head_num + sync_fetch_span)
             return true;
+
+         // might be in irreversible mode
+         controller& cc = my_impl->chain_plug->chain();
+         if (cc.get_read_mode() == db_read_mode::IRREVERSIBLE) {
+            // chain head == lib == fork_db_root in irreversible
+            auto forkdb_head = cc.fork_db_head();
+            if (forkdb_head.block_num() > head_num && forkdb_head.block_num() - head_num >= sync_fetch_span) {
```

Comment:
why do we need `&& forkdb_head.block_num() - head_num >= sync_fetch_span`?

If lib doesn't advance, i.e. `calculated_lib == head_num` below, wouldn't this be already covered by the test above at line 2193?

Reply:
Line 2193 is comparing `head_num` which is the same as `lib` and `fork_db_root` in irreversible mode. This is comparing the last block in the forkdb on the best head to what has been applied. If that is not greater than `sync_fetch_span` then we don't want more blocks.

Reply:
Isn't `blk_num` the last block we received, so similar to `forkdb_head`?

If so, `blk_num < head_num + sync_fetch_span` is similar to `forkdb_head.block_num() < head_num + sync_fetch_span`